src/Http/Routing/src/EndpointSelectorContext.cs

```diff
@@ -3,68 +3,52 @@
 namespace Microsoft.AspNetCore.Routing
 {
-    public sealed class EndpointSelectorContext : IEndpointFeature, IRouteValuesFeature, IRoutingFeature
+    public struct EndpointSelectorContext
     {
-        private RouteData _routeData;
-        private RouteValueDictionary _routeValues;
+        private HttpContext _httpContext;
 
-        /// <summary>
-        /// Gets or sets the selected <see cref="Http.Endpoint"/> for the current
-        /// request.
-        /// </summary>
-        public Endpoint Endpoint { get; set; }
+        public EndpointSelectorContext(HttpContext httpContext)
+        {
+            _httpContext = httpContext;
+        }
 
         /// <summary>
-        /// Gets or sets the <see cref="RouteValueDictionary"/> associated with the currrent
+        /// Gets or sets the selected <see cref="Http.Endpoint"/> for the current
         /// request.
         /// </summary>
-        public RouteValueDictionary RouteValues
+        public Endpoint Endpoint
         {
-            get => _routeValues ?? (_routeValues = new RouteValueDictionary());
+            get
+            {
+                return _httpContext.GetEndpoint();
+            }
             set
             {
-                _routeValues = value;
-
-                // RouteData will be created next get with new Values
-                _routeData = null;
+                if (value != null)
+                {
+                    _httpContext.SetEndpoint(value);
+                }
             }
         }
 
         /// <summary>
-        /// Gets or sets the <see cref="RouteData"/> for the current request.
+        /// Gets or sets the <see cref="RouteValueDictionary"/> associated with the currrent
+        /// request.
         /// </summary>
-        /// <remarks>
-        /// The setter is not implemented. Use <see cref="RouteValues"/> to set the route values.
-        /// </remarks>
-        RouteData IRoutingFeature.RouteData
+        public RouteValueDictionary RouteValues
         {
             get
             {
-                if (_routeData == null)
-                {
-                    _routeData = _routeValues == null ? new RouteData() : new RouteData(_routeValues);
-
-                    // Note: DataTokens won't update if someone else overwrites the Endpoint
-                    // after route values has been set. This seems find since endpoints are a new
-                    // feature and DataTokens are for back-compat.
-                    var dataTokensMetadata = Endpoint?.Metadata.GetMetadata<IDataTokensMetadata>();
-                    if (dataTokensMetadata != null)
-                    {
-                        var dataTokens = _routeData.DataTokens;
-                        foreach (var kvp in dataTokensMetadata.DataTokens)
-                        {
-                            _routeData.DataTokens.Add(kvp.Key, kvp.Value);
-                        }
-                    }
-                }
-
-                return _routeData;
```

Comment:
yeah, looking at this code, I think I'd feel really good about making `RouteData` lazy-init in MVC.